src/patterns.md

```diff
@@ -0,0 +1,646 @@
+# Patterns
+
+> **<sup>Syntax</sup>**\
+> _Pattern_ :\
+> &nbsp;&nbsp; &nbsp;&nbsp; [_LiteralPattern_]\
+> &nbsp;&nbsp; | [_IdentifierPattern_]\
+> &nbsp;&nbsp; | [_WildcardPattern_]\
+> &nbsp;&nbsp; | [_RangePattern_]\
+> &nbsp;&nbsp; | [_ReferencePattern_]\
+> &nbsp;&nbsp; | [_StructPattern_]\
+> &nbsp;&nbsp; | [_TupleStructPattern_]\
+> &nbsp;&nbsp; | [_TuplePattern_]\
+> &nbsp;&nbsp; | [_SlicePattern_]\
+> &nbsp;&nbsp; | [_PathPattern_]
+
+Patterns in Rust are used to match values against structures and to,
+optionally, bind variables to values inside these structures. They are also
+used in variable declarations and function/closure parameters, though in these
+cases most of the time they are simply used as an identifier that binds to a
+value.
+
+For example, the pattern used in:
+
+```rust
+# struct Car;
+# struct Computer;
+# struct Person {
+#     name: String,
+#     car: Option<Car>,
+#     computer: Option<Computer>,
+#     age: u8,
+# }
+# let person = Person {
+#     name: String::from("John"),
+#     car: Some(Car),
+#     computer: None,
+#     age: 15,
+# };
+if let
+    Person {
+        car: Some(_),
+        age: person_age @ 13..=19,
+        name: ref person_name,
+        ..
+    } = person
+{
+    println!("{} has a car and is {} years old.", person_name, person_age);
+}
+```
+does four things:
+
+* Tests if `person` has the `car` field filled with something.
+* Tests if the person's `age` field is between 13 and 19, and binds its value to
+  the `person_age` variable.
+* Binds a reference to the `name` field to the variable `person_name`.
+* Ignores the rest of the fields of `person`, i.e., they can have any value and
+  are not bound to any variables.
+
+Patterns are used in:
+
+* [`let` declarations](statements.html#let-statements)
+* [Function](items/functions.html) and [closure](expressions/closure-expr.html)
+  parameters
+* [`match` expressions](expressions/match-expr.html)
+* [`if let` expressions](expressions/if-expr.html)
+* [`while let` expressions](expressions/loop-expr.html#predicate-pattern-loops)
+* [`for` expressions](expressions/loop-expr.html#iterator-loops)
+* Inside other patterns
```

Comment:
This is a weird way of saying that patterns are recursive. I wouldn't really call a subpattern a pattern on its own right personally.